polynote-kernel/src/main/scala/polynote/messages/Messages.scala

```diff
@@ -127,47 +125,175 @@ object RunCell extends MessageCompanion[RunCell](3)
-  */
-final case class ContentEdit(pos: Int, deleteLength: Int, content: String) {
-  def applyTo(rope: Rope): Rope = if (deleteLength > 0) rope.delete(pos, deleteLength).insertAt(pos, Rope(content)) else rope.insertAt(pos, Rope(content))
 
-  // Given another edit which occured after this edit, create an edit that will be equivalent to applying this edit
-  // when applied after the given edit instead.
-  def rebase(other: ContentEdit): ContentEdit = other match {
-    // if the other edit is entirely before this edit, just shift this edit by the length delta
-    case ContentEdit(otherPos, otherDeleteLength, otherContent) if otherPos + otherDeleteLength <= pos =>
-      copy(pos = pos + otherContent.length - otherDeleteLength)
+sealed trait ContentEdit {
+  def pos: Int
+  def lengthDelta: Int
 
-    // if the other edit is entirely after this edit, nothing to do
-    case ContentEdit(otherPos, _, _) if otherPos >= pos + deleteLength => this
+  def rebase(other: ContentEdits): ContentEdits = ContentEdits(ShortList(ContentEdit.rebaseAll(this, other.edits)._1))
 
-    // if the other edit affects the region that was replaced by this edit, then they conflict
-    // we want the minimal edit that replaces the entire conflict region
-    case ContentEdit(otherPos, otherDeleteLength, otherContent) =>
+  def applyTo(rope: Rope): Rope = this match {
+    case Insert(pos, content) => rope.insertAt(pos, Rope(content))
+    case Delete(pos, length)  => rope.delete(pos, length)
+  }
 
-      // start at the earlier position
-      val start = math.min(pos, otherPos)
+  def nonEmpty: Boolean
+}
 
-      // They deleted to this position
-      val theirTarget = otherPos + otherDeleteLength
+final case class Insert(pos: Int, content: String) extends ContentEdit {
+  def lengthDelta: Int = content.length
 
-      // I would have deleted up to this position...
-      val myTarget = pos + deleteLength
+  def nonEmpty: Boolean = content.nonEmpty
 
-      // So I should delete this much extra
-      val deleteExtra = math.max(0, myTarget - theirTarget) + math.max(0, otherPos - pos)
+  override def toString: String = s"""Insert($pos, "$content")"""
+}
 
-      val combinedContent = if (otherPos <= pos) otherContent + content else content + otherContent
+object Insert {
+  implicit val insertTag: Discriminator[ContentEdit, Insert, Byte] = Discriminator(0)
+}
+
+final case class Delete(pos: Int, length: Int) extends ContentEdit {
+  override def lengthDelta: Int = -length
+
+  override def nonEmpty: Boolean = length != 0
+}
 
-      ContentEdit(start, otherContent.length + deleteExtra, combinedContent)
+object Delete {
+  implicit val deleteTag: Discriminator[ContentEdit, Delete, Byte] = Discriminator(1)
+}
+
+object ContentEdit {
+  implicit val discriminated: Discriminated[ContentEdit, Byte] = Discriminated(byte)
+
+  // rebase a onto b and b onto a
+  def rebase(a: ContentEdit, b: ContentEdit): (List[ContentEdit], List[ContentEdit]) = (a, b) match {
```

Comment:
There are many pairs of cases in this method that duplicate their logic. That's fine, but it did leave me thinking about whether there was a way to avoid the duplication. 

One thing we could do would be to make `rebase` "left-biased" (or "a-biased") by only explicitly handling cases where `a` "wins", and then have a default case that swaps a and b and tries again: 

```scala
    // let's try again
    case (a, b) => rebase(b, a).swap
```

Pro: less duplication
Con: less obvious what's going on, might be hard to reason about?

WDYT?

Here's the full change I'm talking about. 
[diff.txt](https://github.com/polynote/polynote/files/2858380/diff.txt)

Reply:
I thought about that too, but ultimately I think it's clearer with the duplication. It also preserves exhaustivity checking in the compiler, which you'd lose with a fall-through case like that.

Reply:
Makes sense. Compiler checking is a good point I wasn't thinking about!